[NPM] Optimize memory allocation in TCP Failures path

akarpz opened the pull request:
<!--
* New contributors are highly encouraged to read our
  [CONTRIBUTING](/CONTRIBUTING.md) documentation.
* Both Contributor and Reviewer Checklists are available at https://datadoghq.dev/datadog-agent/guidelines/contributing/#pull-requests.
* The pull request:
  * Should only fix one issue or add one feature at a time.
  * Must update the test suite for the relevant functionality.
  * Should pass all status checks before being reviewed or merged.
* Commit titles should be prefixed with general area of pull request's change.
-->
### What does this PR do?

There were a few issues with the existing impl:
- the size of the failures map was uncapped
- failures matched to incoming closed connections were not being immediately removed from the map
- we had an unoptimized allocation pattern for failure stats objects

This PR solves these three issues which should help to resolve increased memory usage on a small number of hosts that were likely seeing large spikes in TCP failures

<!--
* A brief description of the change being made with this pull request.
* If the description here cannot be expressed in a succinct form, consider
  opening multiple pull requests instead of a single one.
-->

### Motivation

<!--
* What inspired you to submit this pull request?
* Link any related GitHub issues or PRs here.
-->

### Additional Notes

in the tcp load test we see significantly fewer items on heap for the methods in question:

link: https://dddev.datadoghq.com/profiling/comparison?compare_end_A=1722484004715&compare_end_B=1722484002521&compare_query_A=service%3Asystem-probe%20agent-env%3Aadamk-loadtest-base%20&compare_query_B=service%3Asystem-probe%20agent-env%3Aadamk-loadtest-new%20&compare_start_A=1722483104715&compare_start_B=1722483102521&compareValuesMode=absolute&comparisonViz=table&group_by=directory&my_code=enabled&profile_type=heap-live-samples

<img width="1229" alt="Screenshot 2024-07-31 at 11 52 45 PM" src="https://github.com/user-attachments/assets/cf67e1f1-5fff-44bf-a576-89f81ef7b265">


It also significantly reduced the incidence of orphaned failed connections as expected:
<img width="436" alt="Screenshot 2024-08-01 at 9 44 43 AM" src="https://github.com/user-attachments/assets/b967b81c-d809-41f7-bd5d-01ac0846faba">
<img width="439" alt="Screenshot 2024-08-01 at 9 51 46 AM" src="https://github.com/user-attachments/assets/09448788-ee15-42c2-9896-ba17dc695cd0">





<!--
* Anything else we should know when reviewing?
* Include benchmarking information here whenever possible.
* Include info about alternatives that were considered and why the proposed
  version was chosen.
-->

### Possible Drawbacks / Trade-offs

<!--
* What are the possible side-effects or negative impacts of the code change?
-->

### Describe how to test/QA your changes

<!--
* Write here in detail or link to detailed instructions on how this change can
  be tested/QAd/validated, including any environment setup.
-->

akarpz committed:
optimize memory allocation in failure path

akarpz committed:
add new config for max failures buffered

akarpz committed:
rename stats pool

akarpz committed:
move comment

akarpz committed:
remove double check for allowed failures. this is done in eBPF now

akarpz committed:
fix comment

brycekahle commented on pkg/network/tracer/connection/failure/matching.go:
```diff
@@ -49,6 +46,12 @@ type FailedConnStats struct {
 }
 
+func (t FailedConnStats) reset() {
+	for k := range t.CountByErrCode {
```
Note: this will not cause the memory usage of the map to be reclaimed. https://github.com/golang/go/issues/20135

brycekahle commented on pkg/network/tracer/connection/failure/matching.go:
```diff
@@ -89,9 +101,8 @@ func (fc *FailedConns) upsertConn(failedConn *ebpf.FailedConn) {
-			CountByErrCode: make(map[uint32]uint32),
-		}
+		stats = fc.statsPool.Get()
+		stats.reset()
```
If you are expecting reset to free up some memory, it would be better to reset before putting back into the pool.

brycekahle commented on pkg/network/tracer/connection/failure/matching.go:
```diff
@@ -110,17 +121,26 @@ func (fc *FailedConns) MatchFailedConn(conn *network.ConnectionStats) {
 	if failedConn, ok := fc.FailedConnMap[*fc.failureTuple]; ok {
 		// found matching failed connection
+		foundMatch = true
 		conn.TCPFailures = make(map[uint32]uint32)
```
It may be better to transfer ownership of the map to the connection, setting `failedConn.CountByErrCode = nil`, and then letting `failedConn` getting GCed. Definitely worth profiling, but if you need to allocate a new map anyways, I'm not sure a `sync.Pool` is helpful.

akarpz replied:
Bryce and I talked offline but we decided to go with this to avoid the extra map allocation

akarpz committed:
get rid of sync pool, assign failure map directly to conn map and delete failure entry on match

akarpz committed:
remove unused method

akarpz committed:
assign map entry to nil after passing it to connection object

hmahmood commented on pkg/network/tracer/connection/failure/matching.go:
```diff
@@ -110,16 +110,23 @@ func (fc *FailedConns) MatchFailedConn(conn *network.ConnectionStats) {
+		failedConn.CountByErrCode = nil
+	}
+	fc.RUnlock()
+	if foundMatch {
```
Why not just delete it above?

brycekahle replied:
Only the read lock is held above.

akarpz replied:
I don't have a perfect understanding of RXMutex, but I was wary about trying to acquire a write lock inside of a read lock

brycekahle replied:
You cannot acquire a write lock while a read lock is held.
> A RWMutex is a reader/writer mutual exclusion lock. The lock can be held by an arbitrary number of readers or a single writer.

hmahmood replied:
I think you may need to use the write lock here. In theory you could have two callers come through with a match on the failures and would have the same failures attached to two connections.

akarpz replied:
that's a fair edge case... I'll just apply a write lock for the entire block

akarpz committed:
use write lock for matching

guyarb commented:
> This PR solves these three issues which should help to resolve increased memory usage on a small number of hosts that were likely seeing large spikes in TCP failures

I'd love to see benchmarks in such a pr, load test and dogfooding to support this

guyarb commented on cmd/system-probe/config/adjust_npm.go:
```diff
@@ -65,6 +65,14 @@ func adjustNetwork(cfg config.Config) {
+	// also ensure that max_failed_connections_buffered is equal to max_tracked_connections if the former is not set
+	validateInt64(cfg, spNS("max_failed_connections_buffered"), cfg.GetInt64(spNS("max_tracked_connections")), func(v int64) error {
+		if v <= 0 {
+			return fmt.Errorf("must be a positive value")
```
nit - 
```suggestion
			return errors.New("must be a positive value")
```

akarpz replied:
done, thanks

guyarb commented on pkg/config/setup/system_probe.go:
```diff
@@ -181,6 +181,7 @@ func InitSystemProbeConfig(cfg pkgconfigmodel.Config) {
 
 	cfg.BindEnvAndSetDefault(join(spNS, "max_tracked_connections"), 65536)
 	cfg.BindEnv(join(spNS, "max_closed_connections_buffered"))
+	cfg.BindEnv(join(spNS, "max_failed_connections_buffered"))
```
why those values are under system-probe configuration and not network_config?

akarpz replied:
I'm happy to move these. this config is very similar to the existing one for closed connections, which is why I put them together

akarpz replied:
I moved this config to `netNS`

akarpz commented:
@guyarb thanks for your review, I added some load test results/profile in the PR description. I modified the tcp-load-test slightly to generate more failures and stress test this particular code path

akarpz committed:
move configs to network ns, use errors.new

derekwbrown reviewed (approved):
approved for (1) test file owned by wkit

brycekahle commented on pkg/network/tracer/connection/failure/matching.go:
```diff
@@ -101,25 +101,22 @@ func (fc *FailedConns) upsertConn(failedConn *ebpf.FailedConn) {
+	if fc == nil || conn.Type != network.TCP {
 		return
 	}
 	util.ConnStatsToTuple(conn, fc.failureTuple)
```
I think this needs to be protected by the mutex

brycekahle commented on pkg/network/tracer/connection/failure/matching.go:
```diff
@@ -61,25 +58,28 @@ type FailedConnMap map[ebpf.ConnTuple]*FailedConnStats
+	maxFailuresBuffered uint32
+	failureTuple        *ebpf.ConnTuple
+	mapCleaner          *ddebpf.MapCleaner[ebpf.ConnTuple, int64]
 	sync.RWMutex
```
This can be a `sync.Mutex` now since you aren't using the read locks.

akarpz committed:
use standard mutex

akarpz committed:
fix locks

brycekahle commented on pkg/network/tracer/connection/failure/matching.go:
```diff
@@ -104,22 +108,24 @@ func (fc *FailedConns) MatchFailedConn(conn *network.ConnectionStats) {
 		return
 	}
+
+	fc.Lock()
```
Probably best to do this after the connection type check, so you don't acquire the lock unnecessarily.

akarpz committed:
reorg checks

akarpz commented:
/merge
